rtic-sync/src/channel/sender.rs

```diff
@@ -0,0 +1,249 @@
+        defmt::write!(f, "Sender",)
+    }
+}
+
+impl<T, const N: usize> Sender<'_, T, N> {
+    /// Try to send a value, non-blocking. If the channel is full this will return an error.
+    pub fn try_send(&mut self, val: T) -> Result<(), TrySendError<T>> {
+        // If the wait queue is not empty, we can't try to push into the queue.
+        // TODO: this no longer seems necessary: freeq items are sent directly to
+        // queueing `send`s.
+        // if !self.0.wait_queue.is_empty() {
+        //     return Err(TrySendError::Full(val));
+        // }
+
+        // No receiver available.
+        if self.is_closed() {
+            return Err(TrySendError::NoReceiver(val));
+        }
+
+        let idx = if let Some(idx) = self.0.pop_free_slot() {
+            idx
+        } else {
+            return Err(TrySendError::Full(val));
+        };
+
+        unsafe { self.0.send_value(idx, val) };
+
+        Ok(())
+    }
+
+    /// Send a value. If there is no place left in the queue this will wait until there is.
+    /// If the receiver does not exist this will return an error.
+    pub async fn send(&mut self, val: T) -> Result<(), NoReceiver<T>> {
+        let mut free_slot_ptr: Option<FreeSlot> = None;
+        let mut link_ptr: Option<Link<WaitQueueData>> = None;
+
+        // Make this future `Drop`-safe.
+        // SAFETY(link_ptr): Shadow the original definition of `link_ptr` so we can't abuse it.
+        let mut link_ptr = LinkPtr(core::ptr::addr_of_mut!(link_ptr));
+        // SAFETY(new): `free_slot_ptr` is alive until at least after `link_ptr` is popped.
+        let mut free_slot_ptr = unsafe { FreeSlotPtr::new(core::ptr::addr_of_mut!(free_slot_ptr)) };
+
+        let mut link_ptr2 = link_ptr.clone();
+        let mut free_slot_ptr2 = free_slot_ptr.clone();
+        let dropper = OnDrop::new(|| {
+            // SAFETY: We only run this closure and dereference the pointer if we have
+            // exited the `poll_fn` below in the `drop(dropper)` call. The other dereference
+            // of this pointer is in the `poll_fn`.
+            if let Some(link) = unsafe { link_ptr2.get() } {
+                self.0.remove_from_wait_queue(link);
+            }
+
+            // Return our potentially-unused free slot.
+            // Potentially unnecessary c-s because our link was already popped, so there
+            // is no way for anything else to access the free slot ptr. Gotta think
+            // about this a bit more...
+            critical_section::with(|cs| {
+                if let Some(freed_slot) = unsafe { free_slot_ptr2.take(cs) } {
+                    // SAFETY: `freed_slot` is a free slot in our referenced channel.
+                    unsafe { self.0.return_free_slot(freed_slot) };
+                }
+            });
+        });
+
+        let idx = poll_fn(|cx| {
+            //  Do all this in one critical section, else there can be race conditions
+            critical_section::with(|cs| {
```

Comment:
Logically, I have reason to believe that this critical section can now, indeed, be removed: all of the critical operations (obtaining a free slot, pushing to the wait queue) are handled atomically by the channel, and are now dependent on each other (a free queue slot is sent to a waiting future directly, so it no longer needs to "snipe" the slot out of the `freeq` independently).

For now, I think sticking them all in a single `critical-section` is the way to go. Once I get `loom` support up and running more, I will definitely investigate whether things do or do not break down if this critical section is removed.